### PR TITLE
extend syntax error information

### PR DESCRIPTION
**Pull request by Simulant87**
The syntax error contains only the index of the not parsable character. It might be helpful to indclude more information like the current character and why it is not allowed at this place.

### CODE REVIEW

**Comment by stleary** [comment]
@Simulant87 Can you include unit tests for the new behavior?

File: src/test/java/org/json/junit/JSONMLTest.java [hunk 1]
```diff
@@ -682,7 +682,7 @@ public void commentsInXML() {
                 "<address>\n"+
                     "<!-- <!--[CDATA[ this is -- <another> comment ]] -->\n"+
                     "<name>Joe Tester</name>\n"+
-                    "<!-- this is a - multi line \n"+
+                    "<!-- this is a - multi in line \n"+
```
Comment (Simulant87) [inline src/test/java/org/json/junit/JSONMLTest.java:685]:
```suggestion
                    "<!-- this is a - multi line \n"+
```

File: src/test/java/org/json/junit/JSONTokenerTest.java [hunk 2]
```diff
@@ -233,82 +233,82 @@ public void testSkipToSuccessWithStringReader() throws IOException {
 
     /**
      * Verify that next and back are working properly and tracking the correct positions
-     * with different new line combinations.
+     * with different new in line combinations.
```
Comment (Simulant87) [inline src/test/java/org/json/junit/JSONTokenerTest.java:236]:
```suggestion
     * with different new line combinations.
```

File: src/test/java/org/json/junit/JSONTokenerTest.java [hunk 3]
```diff
@@ -241,74 +241,74 @@ public void testNextBackComboWithNewLines() {
...
+        assertEquals("skipTo() improperly modifying indexes"," at index 14 [character number 6 in line 2]", tokener.toString());
+        // verify \r\n combo doesn't increment the in line twice
         assertEquals('\r', tokener.next());
-        assertEquals(" at 15 [character 0 line 3]", tokener.toString());
+        assertEquals(" at index 15 [character number 0 in line 3]", tokener.toString());
         assertEquals('\n', tokener.next());
-        assertEquals(" at 16 [character 0 line 3]", tokener.toString());
-        // verify stepping back after reading the \n of an \r\n combo doesn't  increment the line incorrectly
+        assertEquals(" at index 16 [character number 0 in line 3]", tokener.toString());
+        // verify stepping back after reading the \n of an \r\n combo doesn't  increment the in line incorrectly
```
Comment (Simulant87) [inline src/test/java/org/json/junit/JSONTokenerTest.java:262]:
```suggestion
        // verify stepping back after reading the \n of an \r\n combo doesn't  increment the line incorrectly
```

File: src/test/java/org/json/junit/JSONTokenerTest.java [hunk 4]
```diff
@@ -241,74 +241,74 @@ public void testNextBackComboWithNewLines() {
...
-        assertEquals(" at 9 [character 1 line 2]", tokener.toString());
+        assertEquals(" at index 9 [character number 1 in line 2]", tokener.toString());
         tokener.back();
-        assertEquals(" at 8 [character 0 line 2]", tokener.toString());
+        assertEquals(" at index 8 [character number 0 in line 2]", tokener.toString());
         tokener.skipTo('\r');
-        assertEquals("skipTo() improperly modifying indexes"," at 14 [character 6 line 2]", tokener.toString());
-        // verify \r\n combo doesn't increment the line twice
+        assertEquals("skipTo() improperly modifying indexes"," at index 14 [character number 6 in line 2]", tokener.toString());
+        // verify \r\n combo doesn't increment the in line twice
```
Comment (Simulant87) [inline src/test/java/org/json/junit/JSONTokenerTest.java:257]:
```suggestion
        // verify \r\n combo doesn't increment the line twice
```

File: src/test/java/org/json/junit/XMLConfigurationTest.java [hunk 5]
```diff
@@ -298,7 +298,7 @@ public void shouldHandleCommentsInXML() {
                 "   <address>\n"+
                 "       <![CDATA[ this is -- <another> comment ]]>\n"+
                 "       <name>Joe Tester</name>\n"+
-                "       <!-- this is a - multi line \n"+
+                "       <!-- this is a - multi in line \n"+
```
Comment (Simulant87) [inline src/test/java/org/json/junit/XMLConfigurationTest.java:301]:
```suggestion
                "       <!-- this is a - multi line \n"+
```

File: src/test/java/org/json/junit/XMLTest.java [hunk 6]
```diff
@@ -350,7 +350,7 @@ public void shouldHandleCommentsInXML() {
                 "   <address>\n"+
                 "       <![CDATA[ this is -- <another> comment ]]>\n"+
                 "       <name>Joe Tester</name>\n"+
-                "       <!-- this is a - multi line \n"+
+                "       <!-- this is a - multi in line \n"+
```
Comment (Simulant87) [inline src/test/java/org/json/junit/XMLTest.java:353]:
```suggestion
                "       <!-- this is a - multi line \n"+
```

**Comment by Simulant87** [comment]
Sure, I added tests for the new exception messages.
I am not sure, if you like my latest change, explainging the numbers in the syntax error with (zero based) "index" and (1 based) number. I think this makes it easier to find the error when counting for the wrong character in the input string. But it requires a lot of tests to be updated.

If you don't like it, I can alse revert it. But I think it makes the error messages easier to read.

File: src/main/java/org/json/JSONTokener.java [hunk 7]
```diff
@@ -517,11 +520,11 @@ public JSONException syntaxError(String message, Throwable causedBy) {
     /**
      * Make a printable string of this JSONTokener.
      *
-     * @return " at {index} [character {character} line {line}]"
+     * @return " at index {index} [character number {character} in line {line}]"
      */
     @Override
     public String toString() {
-        return " at " + this.index + " [character " + this.character + " line " +
```
Comment (stleary) [inline src/main/java/org/json/JSONTokener.java:524]: Please revert this change. While improving readability is good, I think the benefit is outweighed by the potential cost of broken unit tests by the library users.

Comment (Simulant87) [reply src/main/java/org/json/JSONTokener.java]: reverted.

**Comment by stleary** [comment]
**What problem does this code solve?**
Some tokener parsing error messages are not sufficiently descriptive

**Does the code still compile with Java6?**
Yes

**Risks**
Low. It's possible that some users' unit tests may fail due to this change. I think the value of the new messages outweighs this risk. 

**Changes to the API?**
No, just changes to some parsing error messages.

**Will this require a new release?**
No

**Should the documentation be updated?**
No

**Does it break the unit tests?**
No, there was no previous coverage for these errors.  New tests were added.

**Was any code refactored in this commit?**
No

**Review status**
**APPROVED**

Starting 3-day comment window.

**Review by stleary — APPROVED** [review]
